File: src/java/me/engbers/www/pong/logic/Ball.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package me.engbers.www.pong.logic;

/**
 *
 * @author simon
 */
public class Ball {

    private final int radius;
    private int x;
    private int y;
    private int vx = 10;
    private int vy = 0;
    private final int canvasWidth;
    private final int canvasHeight;

    public Ball(int radius, int x, int y, int vx, int canvasWidth, int canvasHeight) {
        this.radius = radius;
        this.x = x;
        this.y = y;
        this.vx = vx;
        this.canvasHeight = canvasHeight;
        this.canvasWidth = canvasWidth;
    }

    public int getX() {
        return x;
    }

    public void setX(int x) {
        this.x = x;
    }

    public int getY() {
        return y;
    }

    public void setY(int y) {
        this.y = y;
    }

    public int getVx() {
        return vx;
    }

    public int getVy() {
        return vy;
    }

    public void setVy(int vy) {
        this.vy = vy;
    }

    public void setVx(int vx) {
        this.vx = vx;
    }

    public boolean isMovingRight() {
        return vx > 0;
    }

    public boolean isMovingDown() {
        return vy > 0;
    }

    public boolean collidesWith(Player p) {
        if (((y - radius) <= (p.getY() + p.getHeight())) && ((y + radius) >= p.getY())) {
            if (p.getX() == 0) {
                return (x - radius) <= 0;
            } else {
                return (x + radius) >= canvasWidth;
            }
        } else {
            return false;
        }
    }

    public void alternateYSpeed() {
        vy = -vy;
    }

    public void alternateXSpeed() {
        vx = -vx;
    }

    public boolean isTouchingTop() {
        return ((y - radius) <= 0);
    }

    public boolean isTouchingBottom() {
        return ((y + radius) >= canvasHeight);
    }

    public void calculateYSpeed(Player player) {
        if (y <= (player.getY() + (player.getHeight() * 0.25))) {
            vy -= 2;
        } else if (this.y <= (player.getY() + (player.getHeight() * 0.5))) {
            vy -= 1;
        } else if (this.y <= (player.getY() + (player.getHeight() * 0.75))) {
            vy += 1;
        } else {
            vy += 2;
        }
    }
}

File: source/sources/gamelogic/Ball.js
function Ball(radius, x, y, vx, vy, canvasWidth, canvasHeight) {
    this.radius = radius;
    this.x = x;
    this.y = y;
    this.vx = vx;
    this.vy = vy;
    this.canvasWidth = canvasWidth;
    this.canvasHeight = canvasHeight;
}

Ball.prototype.getX = function () {
    return this.x;
};

Ball.prototype.setX = function (x) {
    return this.x = x;
};

Ball.prototype.setY = function (y) {
    return this.y = y;
};

Ball.prototype.getY = function () {
    return this.y;
}

Ball.prototype.getVx = function () {
    return this.vx;
}

Ball.prototype.getVy = function () {
    return this.vy;
}

Ball.prototype.setVx = function (vx) {
    this.vx = vx;
}

Ball.prototype.setVy = function (vy) {
    this.vy = vy;
}

Ball.prototype.getCanvasWidth = function () {
    return this.canvasWidth;
}

Ball.prototype.getCanvasHeight = function () {
    return this.canvasHeight;
}

Ball.prototype.isMovingRight = function () {
    return this.vx > 0;
}
Ball.prototype.isMovingDown = function () {
    return this.vy > 0;
};

Ball.prototype.collidesWith = function (player) {
    if (((this.y - this.radius) <= (player.getY() + player.getHeight())) && ((this.y + this.radius) >= player.getY())) {
        if (player.getX() == 0) {
            return (this.x - this.radius) <= 0;
        } else {
            return (this.x + this.radius) >= this.canvasWidth;
        }
    } else {
        return false;
    }
}

Ball.prototype.alternateXSpeed = function () {
    this.vx = -this.vx;
}

Ball.prototype.alternateYSpeed = function () {
    this.vy = -this.vy;
}

Ball.prototype.isTouchingTop = function () {
    return ((this.y - this.radius) <= 0);
}

Ball.prototype.isTouchingBottom = function () {
    return ((this.y + this.radius) >= this.canvasHeight);
}

Ball.prototype.calculateYSpeed = function (player) {
    if (this.y <= (player.getY() + (player.getHeight() * 0.25))) {
        this.vy -= 2;
    } else if (this.y <= (player.getY() + (player.getHeight() * 0.5))) {
        this.vy -= 1;
    } else if (this.y <= (player.getY() + (player.getHeight() * 0.75))) {
        this.vy += 1;
    } else {
        this.vy += 2;
    }
}


module.exports = Ball;